agent/consul/state/catalog_events.go

```diff
@@ -0,0 +1,465 @@
+package state
+
+import (
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+type changeOp int
+
+const (
+	OpDelete changeOp = iota
+	OpCreate
+	OpUpdate
+)
+
+type eventPayload struct {
+	Op  changeOp
+	Obj interface{}
+}
+
+// serviceHealthSnapshot returns a stream.SnapshotFunc that provides a snapshot
+// of stream.Events that describe the current state of a service health query.
+//
+// TODO: no tests for this yet
+func serviceHealthSnapshot(s *Store, topic topic) stream.SnapshotFunc {
+	return func(req stream.SubscribeRequest, buf stream.SnapshotAppender) (index uint64, err error) {
+		tx := s.db.Txn(false)
+		defer tx.Abort()
+
+		connect := topic == TopicServiceHealthConnect
+		// TODO(namespace-streaming): plumb entMeta through from SubscribeRequest
+		idx, nodes, err := checkServiceNodesTxn(tx, nil, req.Key, connect, nil)
+		if err != nil {
+			return 0, err
+		}
+
+		for _, n := range nodes {
+			event := stream.Event{
+				Index: idx,
+				Topic: topic,
+				Payload: eventPayload{
+					Op:  OpCreate,
+					Obj: &n,
+				},
+			}
+
+			if n.Service != nil {
+				event.Key = n.Service.Service
+			}
+
+			// TODO: could all the events be appended as a single item?
```

Comment:
Actually never mind, it doesn't break invariants because these are snapshot events which all have the same snapshot index already and are not "resumable" (a client can't receive half a snapshot, disconnect and only read the other half on reconnect, they must start again).

We _could_ put some number of these into one event which for now has a possible extremly minor reduction in bytes on the wire from the outer event wrapping. A good motivation to do this would be if/when we implement batching and compressing batches to reduce bandwith. But even then I guess we'd want to do that in one place in the lib not in every snapshot handler so leaving these separate is probably the best thing.

If we did combine we'd need to not put too many in each message otherwise we will affect performance - if there are 10k instances for a service snapshot then we don't really want to create one massive array in memory, encode it as one massive protobuf and send down the wire in one go. Its much better to stream it in small parts.

My vote would be comment why these are not batched for now (i.e. this comment distilled a bit!) and leave them as separate.

Reply:
Makes sense, I'll add a comment to replace the TODO